lib/compress.js

```diff
@@ -4307,6 +4307,17 @@ merge(Compressor.prototype, {
         return self;
     });
 
+    AST_Lambda.DEFMETHOD("contains_this", function() {
+        var result;
+        var self = this;
+        self.walk(new TreeWalker(function(node) {
+            if (result) return true;
+            if (node instanceof AST_This) return result = true;
+            if (node !== self && node instanceof AST_Scope) return true;
+        }));
+        return result;
+    });
```

Comment:
Indeed. Furthermore in the harmony branch an additional optimization can be done to transform both `p(){}` and `p:()=>{}`. Might the `this` issue have to be taken into account for those constructs?